candidate-agreement/src/bft/accumulator.rs

```diff
@@ -0,0 +1,586 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Message accumulator for each round of BFT consensus.
+
+use std::collections::{HashMap, HashSet};
+use std::collections::hash_map::Entry;
+use std::hash::Hash;
+
+use super::{Message, LocalizedMessage};
+
+/// Context necessary to execute a round of BFT.
+pub trait Context {
+	/// A full candidate.
+	type Candidate: Clone;
+	/// Unique digest of a proposed candidate (think hash).
+	type Digest: Hash + Eq + Clone;
+	/// Validator ID.
+	type ValidatorId: Hash + Eq;
+	/// A signature.
+	type Signature: Eq + Clone;
+}
+
+/// Justification at a given round.
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct Justification<D, S> {
+	/// The round.
+	pub round_number: usize,
+	/// The digest prepared for.
+	pub digest: D,
+	/// Signatures for the prepare messages.
+	pub signatures: Vec<S>,
+}
+
+impl<D, S> Justification<D, S> {
+	/// Fails if there are duplicate signatures or invalid.
+	///
+	/// Provide a closure for checking whether the signature is valid on a
+	/// digest.
+	///
+	/// The closure should return true iff the round number, digest, and signature
+	/// represent a valid prepare message and the signer was authorized to issue
+	/// it.
+	pub fn check<F, V>(&self, threshold: usize, check_message: F) -> bool
```

Comment:
reading code where most types are now single-letters (`F`, `D`, `S`, `V`) is pretty hard. as long as we have generic-heavy code like this, we should consider dropping the C++ convention Rust seems to have inherited with naming generics as a single capital letter and go with an actual descriptive name (e.g. `MessageChecker` rather than `F`).